## Premise 
This is the first analysis done by me on the titanic dataset.  
I will go step by step focussing on the pointers provided to me by the vlammrs document related to this assignment.  
The dataset used for the analysis is/or should be stored in the current folder of this script.  

In [1]:
import pandas as pd

In [2]:
help(pd)

Help on package pandas:

NAME
    pandas

DESCRIPTION
    pandas - a powerful data analysis and manipulation library for Python
    
    **pandas** is a Python package providing fast, flexible, and expressive data
    structures designed to make working with "relational" or "labeled" data both
    easy and intuitive. It aims to be the fundamental high-level building block for
    doing practical, **real world** data analysis in Python. Additionally, it has
    the broader goal of becoming **the most powerful and flexible open source data
    analysis / manipulation tool available in any language**. It is already well on
    its way toward this goal.
    
    Main Features
    -------------
    Here are just a few of the things that pandas does well:
    
      - Easy handling of missing data in floating point as well as non-floating
        point data.
      - Size mutability: columns can be inserted and deleted from DataFrame and
        higher dimensional objects
      - Automatic an

In [3]:
dir(pd)


['BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Float64Index',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int64Index',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt64Index',
 'UInt8Dtype',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_hashtable',
 '_is_numpy_dev',
 '_lib',
 '_libs',
 '_np_version_under1p16',
 '_np_version_under1p17',
 '_np_version_under1p18',
 '_testing',
 '_tslib',
 '_ty

In [4]:
titanic = pd.read_excel(r'Titanic_data.xlsx')

In [5]:
titanic.head()

PassengerId Survived2  Pclass  \
0            1       Yes       3   
1            2         0       1   
2            3         0       3   
3            4         0       1   
4            5       Yes       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

             Ticket    Fare Cabin    Embarked  
0         A/5 21171     725   NaN   Cherbourg  
1          PC 17599  712833   C85  Queenstown  
2  STON/O2. 3101282    7925   NaN   Cherbourg  
3            113803     531  C123   Cherbourg  
4            373450     805   NaN   Cherbourg

In [6]:
titanic.shape

(891, 11)

There are 11 columns and data points for 891 passengers. Were these all the passengers on the ship?
***A quick google search shows that 2208 people were onboard, including crew members***  
***1317 passengers 324 in first class, 284 in second class, and 709 in third class*** - source:wiki  
**Of which 1503, or 1512 people died**  
How come we have only data for these members?  
Seems to be that, this is only a test set. Additionally, it is also known that certain orgiinal available has more columns present.  

The question is more on how this dataset was generated?  

Even though there seems to be variety of information, I am not entirely sure which direction to go towards. Should I investigate, what were the causes of the sinkage? Should I investigate what could have been predicted? And what could be avoided, and planned well from now on?

In [7]:
titanic.describe()

PassengerId      Pclass         Age       SibSp          Fare
count   891.000000  891.000000  714.000000  891.000000  8.910000e+02
mean    446.000000    2.308642   38.420168    0.523008  1.276607e+05
std     257.353842    0.836071   56.525211    1.102743  4.111230e+05
min       1.000000    1.000000    1.000000    0.000000  0.000000e+00
25%     223.500000    2.000000   21.000000    0.000000  1.610000e+02
50%     446.000000    3.000000   29.000000    0.000000  2.625000e+03
75%     668.500000    3.000000   40.000000    1.000000  7.895800e+04
max     891.000000    3.000000  705.000000    8.000000  5.123292e+06

In [10]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived2    891 non-null    object 
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    int64  
 9   Cabin        204 non-null    object 
 10  Embarked     891 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 76.7+ KB


**Possible connections - 45**   
1. Class - Fare, Cabin.
>Were people in higher classes paying more than the rest? Should be obvious.
2. Survied to (calss, sex, age, sibsp, Cabin, embarked)
>wealthy people priotised over poorer people. also reads as social class. there higher classes would be given higher priority.  
>Maybe one of the genders is given higher preference.
>Maybe kids and elderly, or a particular age group is given precedence.
>Did kids with more or less siblings survive better than others?
>Cabin : Did location play an influence in the surivival chance of the passengers?
>Were people boarding from certain regions more likely to survive than the rest?
>Does the ticket, and fare play any role in the chances of survival?
>How much influence does each of the attributes play a role in the survival rate of the passengers? 
3. What is the gender distribution in each of the classes?
4. More trivial information such as Name and class?
>Certain surnames more prominent in upper classes?
5. Was there a relationship between SibSp and class?
>Such as wealthier or higher class having more kids? or even the other way round.
6. Was there a relationship between Sex and class?
7. And maybe even between class and embarked.

**Additional curiosities**
1. Was the cabin number missing for a particular class of passengers?
2. Ensuring to remove the bias occuring from Sibsp.

**Quality of Data**
1. Some of the column headers are not very clear, but can be guessed on what they intend.
>survived2 should be survived.  
>Pclass should be passenger class  
>SibSp should be siblings of the passenger. This part was the most difficult part, but can be shown by looking the surnames, age, and SibSp count.   
2. Cabin has a lot of missing values. It only has 204 non-null values.
3. There are also missing values in age column, however it is only for 704 for the total count.
>What ought to be done with the missing values? In case where there is a lot of missing values, it would be most practical drop that particular column. 
>However what should be approach for Age column,? Should I drop it as well, or all the rows for each the value doesn't exist? Or should I fill it with 0, or mean, or using the age distribution. The last one seems to be most solid, mathematically. Since statistical methods are aggregatory, it ought to be fine with the distribution as long as the distribution is maintained. Can I prove it mathematically?  
4. Presence of outliers/anomalies, A quick glimplse earlier showed possibility of anomalies in the data apart from the missing values. Is this true, and can these anomalies be explained, or are they errors during the data creation. What should be done with them incase found.? How to find them?  



In [9]:
titanic_cabinless = titanic.drop(labels='Cabin', axis=1)

In [10]:
titanic_cabinless.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived2    891 non-null    object 
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    int64  
 9   Embarked     891 non-null    object 
dtypes: float64(1), int64(4), object(5)
memory usage: 69.7+ KB


In [21]:
print(titanic_cabinless[titanic_cabinless.Age.isnull()].Pclass.value_counts())
print(titanic_cabinless[titanic_cabinless.Age.isnull()].Survived2.value_counts())
print(titanic_cabinless[titanic_cabinless.Age.isnull()].Sex.value_counts())
print(titanic_cabinless[titanic_cabinless.Age.isnull()].SibSp.value_counts())
print(titanic_cabinless[titanic_cabinless.Age.isnull()].Fare.value_counts())
print(titanic_cabinless[titanic_cabinless.Age.isnull()].Embarked.value_counts())

3    136
1     30
2     11
Name: Pclass, dtype: int64
Yes    125
0       52
Name: Survived2, dtype: int64
male      124
female     53
Name: Sex, dtype: int64
0    137
1     26
8      7
3      4
2      3
Name: SibSp, dtype: int64
775      20
78958    15
805      14
0         8
6955      7
         ..
355       1
86625     1
50        1
95        1
87125     1
Name: Fare, Length: 73, dtype: int64
Cherbourg       90
Southhampton    49
Queenstown      38
Name: Embarked, dtype: int64


In [36]:
titanic_cabinless.Ticket.value_counts().value_counts()

1    547
2     94
3     21
4     11
7      3
6      3
5      2
Name: Ticket, dtype: int64

In [35]:
print(titanic_cabinless.loc[titanic_cabinless['Ticket']=='1601'].Name)
print(titanic_cabinless.loc[titanic_cabinless['Ticket']=='347082'].Name)
print(titanic_cabinless.loc[titanic_cabinless['Ticket']=='CA. 2343'].Name)
print(titanic_cabinless.loc[titanic_cabinless['Ticket']=='CA 2144'].Name)

74       Bing, Mr. Lee
169      Ling, Mr. Lee
509     Lang, Mr. Fang
643    Foo, Mr. Choong
692       Lam, Mr. Ali
826       Lam, Mr. Len
838    Chip, Mr. Chang
Name: Name, dtype: object
13                           Andersson, Mr. Anders Johan
119                    Andersson, Miss. Ellis Anna Maria
541                 Andersson, Miss. Ingeborg Constanzia
542                    Andersson, Miss. Sigrid Elisabeth
610    Andersson, Mrs. Anders Johan (Alfrida Konstant...
813                   Andersson, Miss. Ebba Iris Alfrida
850              Andersson, Master. Sigvard Harald Elias
Name: Name, dtype: object
159           Sage, Master. Thomas Henry
180         Sage, Miss. Constance Gladys
201                  Sage, Mr. Frederick
324             Sage, Mr. George John Jr
792              Sage, Miss. Stella Anna
846             Sage, Mr. Douglas Bullen
863    Sage, Miss. Dorothy Edith "Dolly"
Name: Name, dtype: object
59          Goodwin, Master. William Frederick
71                  Goodwin,

# Understanding each column well, and writing a short description of each
Example
1. Ticket column:
Groups of people who are travelling together are assigned a single ticket.
*Ingenuine ideas: Since groups are assigned by a single ticket, I can find out how survivability affected each group. Did every family lose someone in the mix? What was the distribution of group sizes in each class?*
>Did it also happen than people who were from the same family also assigned a different tickets? So groups travelling together, but each passenger within the group was assigned a unique ticket.? How do I check this?


# Pointers expected in the solution

1. What connections can you find in these data?  
There are 11 columns, with col1 being the primary key marking each passenger with a userid.
Rest of the columns are survived, class, name, sex, age, *sibsp* is not a clear label for the column, ticket is the identifier for tickets bought, fare doesn't denote the currency, cabin describes the cabin label or number of the passengers, embarked is the source of the journey for the passenger.

Steps to understanding the dataset better.
1. Are there columns which do not add new information? Can they be dropped from our analysis?
2. Are there missing values in the each of the columns?
3. Are there outliers or anomalies in the presented data?


2. When you place the data in the present time, for whom is this knowledge interesting?
3. What would your advice be if you were instructed to analyze this data for an insurance company?